rules/enforce-single-end-return.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * @fileoverview Functions must have a single return at the end of the function.
+ * @author Mitchell Van Der Hoeff <mitchellvanderhoeff@gmail.com>
+ */
+
+'use strict';
+
+module.exports = {
+
+    meta: {
+
+        docs: {
+            recommended: false,
+            type: 'error',
+            description: 'Functions must have a single return at the end of the function.'
+        },
+
+        schema: []
+
+    },
+
+    create: function (context) {
+
+        function inspectFunctionDeclaration(emitted) {
+            var node = emitted.node;
+
+            if (emitted.exit) { return; }
+
+            let body = node.body.body;
+            if (body.length === 0) {
+                return;
+            }
+            for (let statement of body.slice(0, -1)) {
+                if (statement['type'] === 'ReturnStatement') {
+                    context.report({
+                        node: node,
+                        message: 'Return statement not at end of function',
+                    });
+                }
+            }
+            let last_statement = body[body.length - 1];
+            if (last_statement['type'] !== 'ReturnStatement') {
```

Comment:
This technique only catches return statements if they're at the top level inside function.
eg- 
```
function foo() {
    if (...) {
        do {
            return 100;
        } while (...)
    }
}
```
this code will not be flagged by current technique.
a way to solve this without any error duplication is:
- Catch all the `ReturnStatement` nodes in addition to current approach.
- Solium provides a function (http://solium.readthedocs.io/en/latest/developer-guide.html#creating-a-core-rule, see `getParent()`) to know the immediate parent of node. If the current return stm.'s immediate parent is not FunctionDeclaration, this means this return is non-top level. Then you can flag it.

Once solved, please add tests for this case . as well.

Reply:
Good catch - one thing though, it looks like we want to catch `BlockStatement` instead of `ReturnStatement`, because we can't get the parent of a parent of a node in the current set-up.

Reply:
ah yes, my bad!